File: Source/Triggers/SetSessionMotionSmoothingSettings.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
#pragma warning disable CS8509 // The switch expression does not handle all possible values of its input type (it is not exhaustive).

namespace Celeste.Mod.MotionSmoothing.Triggers;

[CustomEntity("MotionSmoothing/SetSessionMotionSmoothingSettings")]

public class SetSessionMotionSmoothingSettings : Trigger
{

    private readonly string Enabled;
    private readonly string Framerate;
    private readonly string SmoothCameraString;
    private readonly string RenderMadelineWithSubpixelPrecision;
    private readonly string SmoothBackground;
    private readonly string SmoothForeground;
    private readonly string HideStretchedLevelEdges;
    private readonly string ObjectSmoothingString;
    private readonly string FramerateIncreaseMethodString;
    private readonly string NastyMode;
    
    private UnlockCameraStrategy SmoothCamera;
    private SmoothingMode ObjectSmoothing;
    private UpdateMode FramerateIncreaseMethod;
    
    public SetSessionMotionSmoothingSettings(EntityData data, Vector2 offset) : base(data, offset)
    {
        Enabled = data.Attr("enabled");
        Framerate = data.Attr("framerate");
        SmoothCameraString = data.Attr("smoothCamera");
        RenderMadelineWithSubpixelPrecision = data.Attr("renderMadelineWithSubpixelPrecision");
        SmoothBackground = data.Attr("smoothBackground");
        SmoothForeground = data.Attr("smoothForeground");
        HideStretchedLevelEdges =  data.Attr("hideStretchedLevelEdges");
        ObjectSmoothingString = data.Attr("objectSmoothing");
        FramerateIncreaseMethodString = data.Attr("framerateIncreaseMethod");
        NastyMode = data.Attr("nastyMode");
    }
    
    public override void OnEnter(Player player)
    {
        base.OnEnter(player);

        if (!MotionSmoothingModule.Settings.AllowMapChanges) return;

        MotionSmoothingModule.Session.MapWantsToForceSetSettings = true;
        
        MotionSmoothingModule.Session.MapSetEnabled = Enabled switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetEnabled,
            "OFF" => false,
            "ON" => true
        };
        
        MotionSmoothingModule.Session.MapSetFrameRate = Framerate switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetFrameRate,
            "60" => 60,
            "120" => 120,
            "180" => 180,
            "240" => 240,
            "300" => 300,
            "360" => 360,
            "420" => 420,
            "480" => 480
        };

        SmoothCamera = SmoothCameraString switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetUnlockCameraStrategy,
            "Fancy" => UnlockCameraStrategy.Hires,
            "Fast" => UnlockCameraStrategy.Unlock,
            "Off" => UnlockCameraStrategy.Off
        };
        MotionSmoothingModule.Session.MapSetUnlockCameraStrategy = SmoothCamera;
        
        MotionSmoothingModule.Session.MapSetRenderMadelineWithSubpixels = RenderMadelineWithSubpixelPrecision switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetRenderMadelineWithSubpixels,
            "OFF" => false,
            "ON" => true
        };
        
        MotionSmoothingModule.Session.MapSetRenderBackgroundHires = SmoothBackground switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetRenderBackgroundHires,
            "OFF" => false,
            "ON" => true
        };
        
        MotionSmoothingModule.Session.MapSetRenderForegroundHires = SmoothForeground switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetRenderForegroundHires,
            "OFF" => false,
            "ON" => true
        };
        
        MotionSmoothingModule.Session.MapSetHideStretchedEdges = HideStretchedLevelEdges switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetHideStretchedEdges,
            "OFF" => false,
            "ON" => true
        };

        ObjectSmoothing = ObjectSmoothingString switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetObjectSmoothing,
            "Extrapolate" => SmoothingMode.Extrapolate,
            "Interpolate" => SmoothingMode.Interpolate,
            "Off" => SmoothingMode.Off
        };
        MotionSmoothingModule.Session.MapSetObjectSmoothing = ObjectSmoothing;
        
        FramerateIncreaseMethod = FramerateIncreaseMethodString switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetFramerateIncreaseMethod,
            "Interval" => UpdateMode.Interval,
            "Dynamic" => UpdateMode.Dynamic
        };
        MotionSmoothingModule.Session.MapSetFramerateIncreaseMethod = FramerateIncreaseMethod;

        MotionSmoothingModule.Session.MapSetNastyMode = NastyMode switch
        {
            "Ignore" => MotionSmoothingModule.Session.MapSetNastyMode,
            "OFF" => false,
            "ON" => true
        };
        
        MotionSmoothingModule.Instance.ApplySettings();
    }
}

File: Loenn/triggers/set_session_motion_smoothing_settings.lua
local SetSessionMotionSmoothingSettings = {}

SetSessionMotionSmoothingSettings.name = "MotionSmoothing/SetSessionMotionSmoothingSettings"

SetSessionMotionSmoothingSettings.placements = {
    name = "set_session_motion_smoothing_settings",
    data = {
        width = 16,
        height = 16,
		enabled = "Ignore",
		framerate = "Ignore",
		smoothCamera = "Ignore",
		renderMadelineWithSubpixelPrecision = "Ignore",
		smoothBackground = "Ignore",
		smoothForeground = "Ignore",
		hideStretchedLevelEdges = "Ignore",
		objectSmoothing = "Ignore",
		framerateIncreaseMethod = "Ignore",
		nastyMode = "Ignore"
    }
}

SetSessionMotionSmoothingSettings.fieldOrder = {
	"x", "y",
	"width", "height",
	"editorColor", "enabled",
	"framerate", "smoothCamera",
	"renderMadelineWithSubpixelPrecision", "smoothBackground",
	"smoothForeground", "hideStretchedLevelEdges",
	"objectSmoothing", "framerateIncreaseMethod",
	"nastyMode"
}

SetSessionMotionSmoothingSettings.fieldInformation = {
	enabled = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	},
	framerate = {
		fieldType = "string",
		options = {
			"Ignore",
			"60",
			"120",
			"180",
			"240",
			"300",
			"360",
			"420",
			"480"
		},
		editable = false
	},
	smoothCamera = {
		fieldType = "string",
		options = {
			"Ignore",
			"Fancy",
			"Fast",
			"Off"
		},
		editable = false
	},
	renderMadelineWithSubpixelPrecision = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	},
	smoothBackground = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	},
	smoothForeground = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	},
	hideStretchedLevelEdges = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	},
	objectSmoothing = {
		fieldType = "string",
		options = {
			"Ignore",
			"Extrapolate",
			"Interpolate",
			"Off"
		},
		editable = false
	},
	framerateIncreaseMethod = {
		fieldType = "string",
		options = {
			"Ignore",
			"Interval",
			"Dynamic"
		},
		editable = false
	},
	nastyMode = {
		fieldType = "string",
		options = {
			"Ignore",
			"OFF",
			"ON"
		},
		editable = false
	}
}

return SetSessionMotionSmoothingSettings